[ORC] Bad codegen for global pointer to iterator

### What happened?

```nim
proc foo() =
  iterator bar() {.closure.} =
    discard
  var g {.global.} = rawProc(bar)

foo()
```
```
nim c test.nim
```
```
.../y/.cache/nim/test2_d/@mtest.nim.c: In function ‘NimMainModule’:
.../.cache/nim/test2_d/@mtest.nim.c:269:9: error: ‘colonenv_’ undeclared (first use in this function)
  269 |         colonenv_ = T100001_;
      |         ^~~~~~~~~
.../.cache/nim/test2_d/@mtest.nim.c:269:9: note: each undeclared identifier is reported only once for each function it appears in
```

### Nim Version

Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-11-17
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 1707bc4a992bc966a7439facb9ee819023c22f77
active boot switches: -d:release


### Additional Information

Compiles and runs fine with `--mm:refc`.

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -593,9 +593,12 @@ proc msgSymChoiceUseQualifier(c: PContext; n: PNode; note = errGenerated) =
 
 template isLocalVarSym(n: PNode): bool =
   n.kind == nkSym and 
-    n.sym.kind in {skVar, skLet} and not 
+    (n.sym.kind in {skVar, skLet} and not 
     ({sfGlobal, sfPure} <= n.sym.flags or
-      sfCompileTime in n.sym.flags)
+      sfCompileTime in n.sym.flags) or
+      n.sym.kind in {skProc, skFunc, skIterator} and 
+      sfGlobal notin n.sym.flags
+      )
   
 proc usesLocalVar(n: PNode): bool =
   for z in 1 ..< n.len:

--- a/tests/global/t3505.nim
+++ b/tests/global/t3505.nim
@@ -4,7 +4,7 @@ action: "reject"
 nimout: '''
 t3505.nim(22, 22) Error: cannot assign local to global variable
 t3505.nim(31, 28) Error: cannot assign local to global variable
-
+t3505.nim(39, 29) Error: cannot assign local to global variable
 
 
 
@@ -31,3 +31,11 @@ proc main() =
   var x {.global.} = initX2(initX(f))
   
 main()
+
+# issue #20866
+proc foo2() =
+  iterator bar() {.closure.} =
+    discard
+  var g {.global.} = rawProc(bar)
+
+foo2()

